File: unity/Assets/Scripts/Multiplayer/Players/PlayerDataResponse.cs
using Michitai.Multiplayer.Errors;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using UnityEngine;

namespace Michitai.Multiplayer.Players
{
    /// <summary>
    /// Response containing a player's data with typed deserialization support.
    /// Uses Unity's JsonUtility for deserialization.
    /// </summary>
    /// <typeparam name="T">The type to deserialize player data into.</typeparam>
    [System.Serializable]
    public class PlayerDataResponse<T> : ApiResponse<EGameDataPlayerGetError> where T : class, new()
    {
        [SerializeField]
        private string data_json;   // Unity mode

        /// <summary>
        /// The type of the player data (for identification purposes).
        /// </summary>
        public string type;

        /// <summary>
        /// The unique ID of the player.
        /// </summary>
        public int player_id;

        /// <summary>
        /// The player's display name.
        /// </summary>
        public string player_name;

        /// <summary>
        /// The deserialized player data object using Unity's JsonUtility.
        /// </summary>
        public T Data
        {
            get
            {
                return JsonUtility.FromJson<T>(data_json);
            }
        }
    }
}

File: java/src/main/java/com/michitai/multiplayer/players/PlayerDataResponse.java
package com.michitai.multiplayer.players;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.michitai.multiplayer.ApiResponse;

/**
 * Response containing a player's data with typed deserialization support.
 * Uses Jackson for deserialization.
 *
 * @param <T> The type to deserialize player data into.
 */
public class PlayerDataResponse<T> extends ApiResponse {
    @JsonProperty("type")
    private String type;

    @JsonProperty("player_id")
    private int playerId;

    @JsonProperty("player_name")
    private String playerName;

    @JsonProperty("data_json")
    private String dataJson;

    @JsonProperty("data")
    private T data;

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public int getPlayerId() {
        return playerId;
    }

    public void setPlayerId(int playerId) {
        this.playerId = playerId;
    }

    public String getPlayerName() {
        return playerName;
    }

    public void setPlayerName(String playerName) {
        this.playerName = playerName;
    }

    public String getDataJson() {
        return dataJson;
    }

    public void setDataJson(String dataJson) {
        this.dataJson = dataJson;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }
}
